util/expensivequery/memory_usage_alarm.go

```diff
@@ -0,0 +1,275 @@
+	}
+	record.lastCheckTime = time.Time{}
+	record.tmpDir = filepath.Join(config.GetGlobalConfig().TempStoragePath, "record")
+	record.lastProfileFileName = make([][]string, 2)
+	// Read last records
+	files, err := ioutil.ReadDir(record.tmpDir)
+	if err != nil {
+		return record
+	}
+	for _, f := range files {
+		name := filepath.Join(record.tmpDir, f.Name())
+		if strings.Contains(f.Name(), "running_sql") {
+			record.lastLogFileName = append(record.lastLogFileName, name)
+		}
+		if strings.Contains(f.Name(), "heap") {
+			record.lastProfileFileName[0] = append(record.lastProfileFileName[0], name)
+		}
+		if strings.Contains(f.Name(), "goroutine") {
+			record.lastProfileFileName[1] = append(record.lastProfileFileName[1], name)
+		}
+	}
+
+	return record
+}
+
+// If Performance.ServerMemoryQuota is set, use ServerMemoryQuota * 80% to check oom risk.
+// If Performance.ServerMemoryQuota is not set, use system total memory usage * 80% to check oom risk.
+func (record *memoryUsageAlarm) alarm4ExcessiveMemUsage(sm util.SessionManager, ctx sessionctx.Context) {
+	var memoryUsage uint64
+	instanceStats := &runtime.MemStats{}
+	if record.isServerMemoryQuotaSetByUser {
+		runtime.ReadMemStats(instanceStats)
+		memoryUsage = instanceStats.HeapAlloc
+	} else {
+		memoryUsage, record.err = memory.MemUsed()
+		if record.err != nil {
+			logutil.BgLogger().Error("get system memory usage fail", zap.Error(record.err))
+			return
+		}
+	}
+
+	// TODO: Consider NextGC to record SQLs.
+	if float64(memoryUsage) > float64(record.serverMemoryQuota)*config.GetGlobalConfig().Performance.MemoryUsageAlarmRatio {
+		// At least ten seconds between two recordings that memory usage is less than threshold (default 80% system memory).
+		// If the memory is still exceeded, only records once.
+		interval := time.Since(record.lastCheckTime)
+		record.lastCheckTime = time.Now()
+		if interval > 10*time.Second {
+			record.doRecord(memoryUsage, sm, ctx)
+		}
+	}
+}
+
+func (record *memoryUsageAlarm) doRecord(memUsage uint64, sm util.SessionManager, ctx sessionctx.Context) {
+	memType := "os memory"
+	if record.isServerMemoryQuotaSetByUser {
+		memType = "instance memory"
+	}
+	logutil.BgLogger().Warn("the TiDB instance now takes a lot of memory, has the risk of OOM",
+		zap.Any("memType", memType),
+		zap.Any("memUsage", memUsage),
+		zap.Any("memQuota", record.serverMemoryQuota),
+	)
+
+	if record.err = disk.CheckAndInitTempDir(); record.err != nil {
+		return
+	}
+	record.recordSQLAndSummaryTable(sm, ctx)
+	record.recordProfile()
+
+	tryRemove := func(filename *[]string) {
+		// Keep the last 5 files
+		for len(*filename) > 5 {
+			err := os.Remove((*filename)[0])
+			if err != nil {
+				logutil.BgLogger().Error("remove temp files failed", zap.Error(err))
+				return
+			}
+			*filename = (*filename)[1:]
+		}
+	}
+	tryRemove(&record.lastLogFileName)
+	for i := range record.lastProfileFileName {
+		tryRemove(&record.lastProfileFileName[i])
+	}
```

Comment:
Can we remove the file by a period of time? Such as 10 minutes, rather than record one profile than delete the last files.

Reply:
I think it is not good to remove the file after the file is created for 10 minutes. If TiDB oom and restart, it will only have 10 minutes to save the files, otherwise, the record files will be removed automatically.  
What's the disadvantage of deleting the last files?

Reply:
If TiDB memory usage is continuously high. The profile file will be stored 10s.